File: src/Samples/FSharpApp/Data/OrderModel.fs
namespace FSharpApp.Data

open System
open System.ComponentModel
open Arcadia.ViewModel

type Product = 
    { ID : int
      Name : string
      UnitPrice : float }

type Inventory = 
    { Products : Product [] }

type OrderItem() = 
    inherit ObservableObject()
    let mutable productId = 0
    let mutable units = 0
    
    member this.ProductId 
        with get () = productId
        and set v = 
            productId <- v
            this.RaisePropertyChanged "ProductId"
    
    member this.Units 
        with get () = units
        and set v = 
            units <- v
            this.RaisePropertyChanged "Units"

type Order() = 
    inherit ObservableObject()
    let mutable id = 0
    let mutable date = DateTime.Now
    let mutable items = BindingList<OrderItem>()
    let mutable tax = 0.
    
    member this.ID 
        with get () = id
        and set v = 
            id <- v
            this.RaisePropertyChanged "ID"
    
    member this.Date 
        with get () = date
        and set v = 
            date <- v
            this.RaisePropertyChanged "Date"
    
    member this.Items 
        with get () = items
        and set v = 
            items <- v
            items.ListChanged.Add(fun _ -> this.RaisePropertyChanged "Items")
            this.RaisePropertyChanged "Items"
    
    member this.Tax 
        with get () = tax
        and set v = 
            tax <- v
            this.RaisePropertyChanged "Tax"

type OrderResult = 
    { TotalUnits : int
      PreTaxAmount : float
      TaxAmount : float
      TotalAmount : float }

File: src/Samples/CSharpApp/Data/OrderModel.cs
namespace CSharpApp.Data
{
    using System;
    using System.ComponentModel;
    using Arcadia.MVVM;

    public class Product
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public double UnitPrice { get; set; }
    }

    public class Inventory
    {
        public Product[] Products { get; set; }
    }

    public class OrderItem : ObservableObject
    {
        int _productId;
        int _units;

        public OrderItem()
        {
            _productId = 0;
            _units = 0;
        }

        public int ProductId
        {
            get { return _productId; }
            set
            {
                _productId = value;
                RaisePropertyChanged("ProductId");
            }
        }

        public int Units
        {
            get { return _units; }
            set
            {
                _units = value;
                RaisePropertyChanged("Units");
            }
        }
    }

    public class Order : ObservableObject
    {
        int _id;
        DateTime _date;
        double _tax;
        BindingList<OrderItem> _items;

        public int ID
        {
            get { return _id; }
            set
            {
                _id = value;
                RaisePropertyChanged("ID");
            }
        }

        public DateTime Date
        {
            get { return _date; }
            set
            {
                _date = value;
                RaisePropertyChanged("Date");
            }
        }

        public BindingList<OrderItem> Items
        {
            get { return _items; }
            set
            {
                _items = value;
                _items.ListChanged += new ListChangedEventHandler((sender, args) => RaisePropertyChanged("Items"));
            }
        }

        public double Tax
        {
            get { return _tax; }
            set
            {
                _tax = value;
                RaisePropertyChanged("Tax");
            }
        }
    }

    public class OrderResult
    {
        public int TotalUnits { get; set; }
        public double PreTaxAmount { get; set; }
        public double TaxAmount { get; set; }
        public double TotalAmount { get; set; }
    }
}